examples/fabric/build_your_own_trainer/trainer.py

```diff
@@ -0,0 +1,522 @@
+        self.current_epoch = 0
+
+        self.max_epochs = max_epochs
+        self.max_steps = max_steps
+        self.should_stop = False
+
+        # ensures limit_X_batches is either int or inf
+        if not isinstance(limit_train_batches, int):
+            assert limit_train_batches == float("inf")
+
+        if not isinstance(limit_val_batches, int):
+            assert limit_val_batches == float("inf")
+
+        self.limit_train_batches = limit_train_batches
+        self.limit_val_batches = limit_val_batches
+        self.validation_frequency = validation_frequency
+        self.use_distributed_sampler = use_distributed_sampler
+        self._current_train_return: Union[torch.Tensor, Mapping[str, Any]] = {}
+        self._current_val_return: Optional[Union[torch.Tensor, Mapping[str, Any]]] = {}
+
+        self.checkpoint_dir = checkpoint_dir
+        self.checkpoint_frequency = checkpoint_frequency
+
+    def fit(
+        self,
+        model: L.LightningModule,
+        train_loader: torch.utils.data.DataLoader,
+        val_loader: torch.utils.data.DataLoader,
+    ):
+        """The main entrypoint of the trainer, triggering the actual training.
+
+        Args:
+            model: the LightningModule to train.
+                Can have the same hooks as :attr:`callbacks` (see :meth:`MyCustomTrainer.__init__`).
+            train_loader: the training dataloader. Has to be an iterable returning batches.
+            val_loader: the validation dataloader. Has to be an iterable returning batches.
+                If not specified, no validation will run.
+        """
+        self.fabric.launch()
+
+        # setup dataloaders
+        train_loader = self.fabric.setup_dataloaders(train_loader, use_distributed_sampler=self.use_distributed_sampler)
+        if val_loader is not None:
+            val_loader = self.fabric.setup_dataloaders(val_loader, use_distributed_sampler=self.use_distributed_sampler)
+
+        # setup model and optimizer
+        if isinstance(self.fabric.strategy, L.fabric.strategies.fsdp.FSDPStrategy):
+            # currently, there is no way to support fsdp with model.configure_optimizers in fabric
+            # as it would require fabric to hold a reference to the model, which we don't want to.
+            raise NotImplementedError("BYOT currently does not support FSDP")
+        else:
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            assert optimizer is not None
+            model, optimizer = self.fabric.setup(model, optimizer)
+
+        # assemble state (current epoch and global step will be added in save)
+        state = {"model": model, "optim": optimizer, "scheduler": scheduler_cfg}
+
+        # load last checkpoint if available
+        latest_checkpoint_path = self.get_latest_checkpoint(self.checkpoint_dir)
+        if latest_checkpoint_path is not None:
+            self.load(state, latest_checkpoint_path)
```

Comment:
I'm not sure this should be part of `fit` if it doesn't include a ckpt_path argument.

This means that this forcefully loads a checkpoint. You might not want to.

I suggest either removing this and doing it before `.fit` or adding `ckpt_path` support

Reply:
I'll address this tomorrow in a small follow-up

Reply:
I'm a bit late on this, but added in #17076